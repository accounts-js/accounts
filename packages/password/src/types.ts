import { CreateUserType, HashAlgorithm, LoginUserIdentityType } from '@accounts/common';

export type PasswordType =
  | string
  | {
      digest: string;
      algorithm: HashAlgorithm;
    };

export interface PasswordLoginType {
  user: string | LoginUserIdentityType;
  // User password
  password: PasswordType;
  // Two factor code
  code?: string;
}

export interface PasswordCreateUserType extends CreateUserType {
  password: PasswordType;
}
