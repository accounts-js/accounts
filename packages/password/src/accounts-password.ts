import { trim, isEmpty, isFunction, isString, isPlainObject, find, includes } from 'lodash';
import {
  CreateUser,
  User,
  LoginUserIdentity,
  EmailRecord,
  TokenRecord,
  DatabaseInterface,
  AuthenticationService,
  HashAlgorithm,
} from '@accounts/types';
import { TwoFactor, AccountsTwoFactorOptions, getUserTwoFactorService } from '@accounts/two-factor';
import { AccountsServer, generateRandomToken, getFirstUserEmail } from '@accounts/server';
import {
  getUserResetTokens,
  getUserVerificationTokens,
  hashPassword,
  bcryptPassword,
  verifyPassword,
  isEmail,
} from './utils';
import { PasswordCreateUserType, PasswordLoginType, PasswordType } from './types';

export interface AccountsPasswordOptions {
  twoFactor?: AccountsTwoFactorOptions;
  passwordHashAlgorithm?: HashAlgorithm;
  /**
   * The number of milliseconds from when a link to verify the user email is sent until token expires and user can't verify his email with the link anymore. Defaults to 3 days.
   */
  verifyEmailTokenExpiration?: number;
  /**
   * The number of milliseconds from when a link to reset password is sent until token expires and user can't reset password with the link anymore. Defaults to 3 days.
   */
  passwordResetTokenExpiration?: number;
  /**
   * The number of milliseconds from when a link to set inital password is sent until token expires and user can't set password with the link anymore. Defaults to 30 days.
   */
  passwordEnrollTokenExpiration?: number;
  minimumPasswordLength?: number;
  validateNewUser?: (user: CreateUser) => Promise<boolean>;
  validateEmail?(email?: string): boolean;
  validatePassword?(password?: PasswordType): boolean;
  validateUsername?(username?: string): boolean;
}

const defaultOptions = {
  minimumPasswordLength: 7,
  // 3 days - 3 * 24 * 60 * 60 * 1000
  verifyEmailTokenExpiration: 259200000,
  // 3 days - 3 * 24 * 60 * 60 * 1000
  passwordResetTokenExpiration: 259200000,
  // 30 days - 30 * 24 * 60 * 60 * 1000
  passwordEnrollTokenExpiration: 2592000000,
  validateEmail(email?: string): boolean {
    const isValid = !isEmpty(trim(email || '')) && isEmail(email);
    return Boolean(isValid);
  },
  validatePassword(password?: PasswordType): boolean {
    const isValid = !isEmpty(password);
    return isValid;
  },
  validateUsername(username?: string): boolean {
    const usernameRegex = /^[a-zA-Z][a-zA-Z0-9]*$/;
    const isValid = username && !isEmpty(trim(username)) && usernameRegex.test(username);
    return Boolean(isValid);
  },
};

export default class AccountsPassword implements AuthenticationService {
  public serviceName = 'password';
  public server!: AccountsServer;
  public twoFactor: TwoFactor;
  private options: AccountsPasswordOptions & typeof defaultOptions;
  private db!: DatabaseInterface;

  constructor(options: AccountsPasswordOptions = {}) {
    this.options = { ...defaultOptions, ...options };
    this.twoFactor = new TwoFactor(options.twoFactor);
  }

  public setStore(store: DatabaseInterface) {
    this.db = store;
    this.twoFactor.setStore(store);
  }

  public async authenticate(params: PasswordLoginType): Promise<User> {
    const { user, password, code } = params;
    if (!user || !password) {
      throw new Error('Unrecognized options for login request');
    }
    if ((!isString(user) && !isPlainObject(user)) || !isString(password)) {
      throw new Error('Match failed');
    }

    const foundUser = await this.passwordAuthenticator(user, password);

    // If user activated two factor authentication try with the code
    if (getUserTwoFactorService(foundUser)) {
      await this.twoFactor.authenticate(foundUser, code!);
    }

    return foundUser;
  }

  /**
   * @description Find a user by one of his emails.
   * @param {string} email - User email.
   * @returns {Promise<Object>} - Return a user or null if not found.
   */
  public findUserByEmail(email: string): Promise<User | null> {
    return this.db.findUserByEmail(email);
  }

  /**
   * @description Find a user by his username.
   * @param {string} username - User username.
   * @returns {Promise<Object>} - Return a user or null if not found.
   */
  public findUserByUsername(username: string): Promise<User | null> {
    return this.db.findUserByUsername(username);
  }

  /**
   * @description Add an email address for a user.
   * Use this instead of directly updating the database.
   * @param {string} userId - User id.
   * @param {string} newEmail - A new email address for the user.
   * @param {boolean} [verified] - Whether the new email address should be marked as verified.
   * Defaults to false.
   * @returns {Promise<void>} - Return a Promise.
   */
  public addEmail(userId: string, newEmail: string, verified: boolean): Promise<void> {
    // TODO use this.options.verifyEmail before
    return this.db.addEmail(userId, newEmail, verified);
  }

  /**
   * @description Remove an email address for a user.
   * Use this instead of directly updating the database.
   * @param {string} userId - User id.
   * @param {string} email - The email address to remove.
   * @returns {Promise<void>} - Return a Promise.
   */
  public removeEmail(userId: string, email: string): Promise<void> {
    return this.db.removeEmail(userId, email);
  }

  /**
   * @description Marks the user's email address as verified.
   * @param {string} token - The token retrieved from the verification URL.
   * @returns {Promise<void>} - Return a Promise.
   */
  public async verifyEmail(token: string): Promise<void> {
    const user = await this.db.findUserByEmailVerificationToken(token);
    if (!user) {
      throw new Error('Verify email link expired');
    }

    const verificationTokens = getUserVerificationTokens(user);
    const tokenRecord = find(verificationTokens, (t: TokenRecord) => t.token === token);
    if (!tokenRecord || this.isTokenExpired(tokenRecord, this.options.verifyEmailTokenExpiration)) {
      throw new Error('Verify email link expired');
    }

    const emailRecord = find(user.emails, (e: EmailRecord) => e.address === tokenRecord.address);
    if (!emailRecord) {
      throw new Error('Verify email link is for unknown address');
    }
    await this.db.verifyEmail(user.id, emailRecord.address);
  }

  /**
   * @description Reset the password for a user using a token received in email.
   * @param {string} token - The token retrieved from the reset password URL.
   * @param {string} newPassword - A new password for the user.
   * @returns {Promise<void>} - Return a Promise.
   */
  public async resetPassword(token: string, newPassword: PasswordType): Promise<void> {
    const user = await this.db.findUserByResetPasswordToken(token);
    if (!user) {
      throw new Error('Reset password link expired');
    }

    const resetTokens = getUserResetTokens(user);
    const resetTokenRecord = find(resetTokens, t => t.token === token);

    if (
      !resetTokenRecord ||
      this.isTokenExpired(
        resetTokenRecord,
        resetTokenRecord.reason === 'enroll'
          ? this.options.passwordEnrollTokenExpiration
          : this.options.passwordResetTokenExpiration
      )
    ) {
      throw new Error('Reset password link expired');
    }

    const emails = user.emails || [];
    if (!includes(emails.map((email: EmailRecord) => email.address), resetTokenRecord.address)) {
      throw new Error('Token has invalid email address');
    }

    const password = await this.hashAndBcryptPassword(newPassword);
    // Change the user password and remove the old token
    await this.db.setResetPassword(user.id, resetTokenRecord.address, password, token);

    // If user clicked on an enrollment link we can verify his email
    if (resetTokenRecord.reason === 'enroll') {
      await this.db.verifyEmail(user.id, resetTokenRecord.address);
    }

    // Changing the password should invalidate existing sessions
    this.db.invalidateAllSessions(user.id);
  }

  /**
   * @description Change the password for a user.
   * @param {string} userId - User id.
   * @param {string} newPassword - A new password for the user.
   * @returns {Promise<void>} - Return a Promise.
   */
  public async setPassword(userId: string, newPassword: string): Promise<void> {
    const password = await bcryptPassword(newPassword);
    return this.db.setPassword(userId, password);
  }

  /**
   * @description Change the current user's password.
   * @param {string} userId - User id.
   * @param {string} oldPassword - The user's current password.
   * @param {string} newPassword - A new password for the user.
   * @returns {Promise<void>} - Return a Promise.
   */
  public async changePassword(
    userId: string,
    oldPassword: string,
    newPassword: string
  ): Promise<void> {
    await this.passwordAuthenticator({ id: userId }, oldPassword);
    const password = await bcryptPassword(newPassword);
    return this.db.setPassword(userId, password);
  }

  /**
   * @description Send an email with a link the user can use verify their email address.
   * @param {string} [address] - Which address of the user's to send the email to.
   * This address must be in the user's emails list.
   * Defaults to the first unverified email in the list.
   * @returns {Promise<void>} - Return a Promise.
   */
  public async sendVerificationEmail(address: string): Promise<void> {
    if (!address) {
      throw new Error('Invalid email');
    }
    const user = await this.db.findUserByEmail(address);
    if (!user) {
      throw new Error('User not found');
    }
    // Make sure the address is valid
    const emails = user.emails || [];
    if (!address || !includes(emails.map(email => email.address), address)) {
      throw new Error('No such email address for user');
    }
    const token = generateRandomToken();
    await this.db.addEmailVerificationToken(user.id, address, token);

    const resetPasswordMail = this.server.prepareMail(
      address,
      token,
      this.server.sanitizeUser(user),
      'verify-email',
      this.server.options.emailTemplates.verifyEmail,
      this.server.options.emailTemplates.from
    );

    await this.server.options.sendMail(resetPasswordMail);
  }

  /**
   * @description Send an email with a link the user can use to reset their password.
   * @param {string} [address] - Which address of the user's to send the email to.
   * This address must be in the user's emails list.
   * Defaults to the first email in the list.
   * @returns {Promise<void>} - Return a Promise.
   */
  public async sendResetPasswordEmail(address: string): Promise<void> {
    if (!address) {
      throw new Error('Invalid email');
    }
    const user = await this.db.findUserByEmail(address);
    if (!user) {
      throw new Error('User not found');
    }
    address = getFirstUserEmail(user, address);
    const token = generateRandomToken();
    await this.db.addResetPasswordToken(user.id, address, token, 'reset');

    const resetPasswordMail = this.server.prepareMail(
      address,
      token,
      this.server.sanitizeUser(user),
      'reset-password',
      this.server.options.emailTemplates.resetPassword,
      this.server.options.emailTemplates.from
    );

    await this.server.options.sendMail(resetPasswordMail);
  }

  /**
   * @description Send an email with a link the user can use to set their initial password.
   * The user's email will be verified after clicking on the link.
   * @param {string} [address] - Which address of the user's to send the email to.
   * This address must be in the user's emails list.
   * Defaults to the first email in the list.
   * @returns {Promise<void>} - Return a Promise.
   */
  public async sendEnrollmentEmail(address: string): Promise<void> {
    const user = await this.db.findUserByEmail(address);
    if (!user) {
      throw new Error('User not found');
    }
    address = getFirstUserEmail(user, address);
    const token = generateRandomToken();
    await this.db.addResetPasswordToken(user.id, address, token, 'enroll');

    const enrollmentMail = this.server.prepareMail(
      address,
      token,
      this.server.sanitizeUser(user),
      'enroll-account',
      this.server.options.emailTemplates.enrollAccount,
      this.server.options.emailTemplates.from
    );

    await this.server.options.sendMail(enrollmentMail);
  }

  /**
   * @description Create a new user.
   * @param user - The user object.
   * @returns Return the id of user created.
   */
  public async createUser(user: PasswordCreateUserType): Promise<string> {
    if (!this.options.validateUsername(user.username) && !this.options.validateEmail(user.email)) {
      throw new Error('Username or Email is required');
    }

    if (user.username && (await this.db.findUserByUsername(user.username))) {
      throw new Error('Username already exists');
    }

    if (user.email && (await this.db.findUserByEmail(user.email))) {
      throw new Error('Email already exists');
    }

    let password;
    if (user.password) {
      if (!this.options.validatePassword(user.password)) {
        throw new Error('Invalid password');
      }
      password = await this.hashAndBcryptPassword(user.password);
    }

    const proposedUserObject = {
      username: user.username,
      email: user.email && user.email.toLowerCase(),
      password,
      profile: user.profile,
    };

    const { validateNewUser } = this.options;
    if (isFunction(validateNewUser) && !(await validateNewUser(proposedUserObject))) {
      throw new Error('User invalid');
    }

    return this.db.createUser(proposedUserObject);
  }

  public isTokenExpired(tokenRecord: TokenRecord, expiryDate: number): boolean {
    return Number(tokenRecord.when) + expiryDate < Date.now();
  }

  private async passwordAuthenticator(
    user: string | LoginUserIdentity,
    password: PasswordType
  ): Promise<User> {
    const { username, email, id } = isString(user)
      ? this.toUsernameAndEmail({ user })
      : this.toUsernameAndEmail({ ...user });

    let foundUser: User | null;

    if (id) {
      // this._validateLoginWithField('id', user);
      foundUser = await this.db.findUserById(id);
    } else if (username) {
      // this._validateLoginWithField('username', user);
      foundUser = await this.db.findUserByUsername(username);
    } else if (email) {
      // this._validateLoginWithField('email', user);
      foundUser = await this.db.findUserByEmail(email);
    }

    // @ts-ignore
    if (!foundUser) {
      throw new Error('User not found');
    }

    const hash = await this.db.findPasswordHash(foundUser.id);
    if (!hash) {
      throw new Error('User has no password set');
    }

    const hashAlgorithm = this.options.passwordHashAlgorithm;
    const pass: any = hashAlgorithm ? hashPassword(password, hashAlgorithm) : password;
    const isPasswordValid = await verifyPassword(pass, hash);

    if (!isPasswordValid) {
      throw new Error('Incorrect password');
    }

    return foundUser;
  }

  private async hashAndBcryptPassword(password: PasswordType): Promise<string> {
    const hashAlgorithm = this.options.passwordHashAlgorithm;
    const hashedPassword: any = hashAlgorithm ? hashPassword(password, hashAlgorithm) : password;
    return bcryptPassword(hashedPassword);
  }

  /**
   * Given a username, user and/or email figure out the username and/or email.
   *
   * @param user An object containing at least `username`, `user` and/or `email`.
   * @returns An object containing `id`, `username` and `email`.
   */
  private toUsernameAndEmail({ user, username, email, id }: any): any {
    if (user && !username && !email) {
      if (isEmail(user)) {
        email = user;
        username = null;
      } else {
        username = user;
        email = null;
      }
    }
    return { username, email, id };
  }
}
