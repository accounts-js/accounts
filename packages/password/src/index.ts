import AccountsPassword from './accounts-password';
export * from './types';

export default AccountsPassword;
export { AccountsPassword };
