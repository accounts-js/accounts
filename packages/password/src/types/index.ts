export { ErrorMessages } from './error-messages';
