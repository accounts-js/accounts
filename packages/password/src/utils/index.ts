export { getUserResetTokens, getUserVerificationTokens } from './user';
export { hashPassword, bcryptPassword, verifyPassword } from './encryption';
export { isEmail } from './is-email';
