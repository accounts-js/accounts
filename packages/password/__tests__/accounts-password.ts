import { set } from 'lodash';
import { AccountsPassword } from '../src';

describe('AccountsPassword', () => {
  const password = new AccountsPassword({});

  describe('config', () => {
    it('should have default options', async () => {
      expect(password.options.passwordEnrollTokenExpiration).toBe(2592000000);
    });
  });

  describe('authenticate', () => {
    it('throws on invalid params', async () => {
      try {
        await password.authenticate({} as any);
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('throws on invalid type params', async () => {
      try {
        await password.authenticate({ user: 'toto', password: 3 } as any);
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('return user', async () => {
      const user = {
        services: {},
      };
      const tmpAccountsPassword = new AccountsPassword({});
      tmpAccountsPassword.passwordAuthenticator = jest.fn(() => Promise.resolve(user));
      const ret = await tmpAccountsPassword.authenticate({
        user: 'toto',
        password: 'toto',
      } as any);
      expect(ret).toEqual(user);
    });

    it('throws when user not found', async () => {
      const findUserByEmail = jest.fn(() => Promise.resolve());
      password.setStore({ findUserByEmail } as any);
      try {
        await password.authenticate({
          user: 'toto@toto.com',
          password: 'toto',
        } as any);
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('throws when hash not found', async () => {
      const findUserByEmail = jest.fn(() => Promise.resolve({ id: 'id' }));
      const findPasswordHash = jest.fn(() => Promise.resolve());
      password.setStore({ findUserByEmail, findPasswordHash } as any);
      try {
        await password.authenticate({
          user: 'toto@toto.com',
          password: 'toto',
        } as any);
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('throws on incorrect password', async () => {
      const findUserByEmail = jest.fn(() => Promise.resolve({ id: 'id' }));
      const findPasswordHash = jest.fn(() => Promise.resolve('hash'));
      password.setStore({ findUserByEmail, findPasswordHash } as any);
      try {
        await password.authenticate({
          user: 'toto@toto.com',
          password: 'toto',
        } as any);
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });
  });

  describe('findUserByEmail', () => {
    it('call this.db.findUserByEmail', async () => {
      const findUserByEmail = jest.fn(() => Promise.resolve('user'));
      password.setStore({ findUserByEmail } as any);
      const user = await password.findUserByEmail('email');
      expect(findUserByEmail.mock.calls[0]).toMatchSnapshot();
      expect(user).toEqual('user');
    });
  });

  describe('findUserByUsername', () => {
    it('call this.db.findUserByUsername', async () => {
      const findUserByUsername = jest.fn(() => Promise.resolve('user'));
      password.setStore({ findUserByUsername } as any);
      const user = await password.findUserByUsername('email');
      expect(findUserByUsername.mock.calls[0]).toMatchSnapshot();
      expect(user).toEqual('user');
    });
  });

  describe('addEmail', () => {
    it('call this.db.addEmail', async () => {
      const addEmail = jest.fn(() => Promise.resolve());
      password.setStore({ addEmail } as any);
      await password.addEmail('id', 'email', true);
      expect(addEmail.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('removeEmail', () => {
    it('call this.db.removeEmail', async () => {
      const removeEmail = jest.fn(() => Promise.resolve());
      password.setStore({ removeEmail } as any);
      await password.removeEmail('id', 'email');
      expect(removeEmail.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('verifyEmail', () => {
    const token = 'token';
    const validUser: any = {};
    const email = 'john.doe@gmail.com';
    set(validUser, 'services.email.verificationTokens', [{ token, address: email }]);
    validUser.emails = [{ address: email }];
    const invalidUser = { ...validUser };
    invalidUser.emails = [];

    it('throws when no user is found', async () => {
      const findUserByEmailVerificationToken = jest.fn(() => Promise.resolve());
      password.setStore({ findUserByEmailVerificationToken } as any);
      try {
        await password.verifyEmail(token);
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('throws when token not found', async () => {
      const isTokenExpired = jest.fn(() => true);
      password.server = { isTokenExpired } as any;
      const findUserByEmailVerificationToken = jest.fn(() => Promise.resolve({}));
      password.setStore({ findUserByEmailVerificationToken } as any);
      try {
        await password.verifyEmail(token);
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('throws when email not found', async () => {
      const isTokenExpired = jest.fn(() => false);
      password.server = { isTokenExpired } as any;
      const findUserByEmailVerificationToken = jest.fn(() => Promise.resolve(invalidUser));
      password.setStore({ findUserByEmailVerificationToken } as any);
      try {
        await password.verifyEmail(token);
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('call this.db.verifyEmail', async () => {
      const isTokenExpired = jest.fn(() => false);
      password.server = { isTokenExpired } as any;
      const findUserByEmailVerificationToken = jest.fn(() => Promise.resolve(validUser));
      const verifyEmail = jest.fn(() => Promise.resolve());
      password.setStore({
        findUserByEmailVerificationToken,
        verifyEmail,
      } as any);
      await password.verifyEmail(token);
      expect(verifyEmail.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('resetPassword', () => {
    const token = 'token';
    const newPassword = 'newPassword';
    const email = 'john.doe@gmail.com';
    const validUser: any = {
      id: 'id',
    };
    set(validUser, 'services.password.reset', [{ token, address: email }]);
    validUser.emails = [{ address: email }];
    const validUserEnroll: any = {
      id: 'id',
    };
    validUserEnroll.emails = [{ address: email }];
    set(validUserEnroll, 'services.password.reset', [{ token, address: email, reason: 'enroll' }]);
    const invalidUser = { ...validUser };
    invalidUser.emails = [];

    it('throws when token not found', async () => {
      const findUserByResetPasswordToken = jest.fn(() => Promise.resolve());
      password.setStore({ findUserByResetPasswordToken } as any);
      try {
        await password.resetPassword(token, newPassword);
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('throws when token is expired', async () => {
      const findUserByResetPasswordToken = jest.fn(() => Promise.resolve(invalidUser));
      password.isTokenExpired = jest.fn(() => true);
      password.setStore({ findUserByResetPasswordToken } as any);
      try {
        await password.resetPassword(token, newPassword);
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('throws when token have invalid email', async () => {
      const findUserByResetPasswordToken = jest.fn(() => Promise.resolve(invalidUser));
      password.isTokenExpired = jest.fn(() => false);
      password.setStore({ findUserByResetPasswordToken } as any);
      try {
        await password.resetPassword(token, newPassword);
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('validate user email if enrolled', async () => {
      const findUserByResetPasswordToken = jest.fn(() => Promise.resolve(validUserEnroll));
      password.isTokenExpired = jest.fn(() => false);
      const setResetPassword = jest.fn(() => Promise.resolve());
      const invalidateAllSessions = jest.fn(() => Promise.resolve());
      const verifyEmail = jest.fn(() => Promise.resolve());
      password.setStore({
        findUserByResetPasswordToken,
        setResetPassword,
        invalidateAllSessions,
        verifyEmail,
      } as any);
      await password.resetPassword(token, newPassword);
      expect(setResetPassword.mock.calls.length).toBe(1);
      expect(verifyEmail.mock.calls.length).toBe(1);
      expect(invalidateAllSessions.mock.calls[0]).toMatchSnapshot();
    });

    it('reset password and invalidate all sessions', async () => {
      const findUserByResetPasswordToken = jest.fn(() => Promise.resolve(validUser));
      const isTokenExpired = jest.fn(() => false);
      const setResetPassword = jest.fn(() => Promise.resolve());
      const invalidateAllSessions = jest.fn(() => Promise.resolve());
      password.setStore({
        findUserByResetPasswordToken,
        setResetPassword,
        invalidateAllSessions,
      } as any);
      password.server = { isTokenExpired } as any;
      await password.resetPassword(token, newPassword);
      expect(setResetPassword.mock.calls.length).toBe(1);
      expect(invalidateAllSessions.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('setPassword', () => {
    it('call this.db.setPassword', async () => {
      const userId = 'id';
      const setPassword = jest.fn(() => Promise.resolve('user'));
      password.setStore({ setPassword } as any);
      const user = await password.setPassword(userId, 'new-password');
      expect(setPassword.mock.calls[0][0]).toEqual(userId);
      expect(setPassword.mock.calls[0][1]).toBeTruthy();
      expect(user).toEqual('user');
    });
  });

  describe('changePassword', () => {
    it('call passwordAuthenticator and this.db.setPassword', async () => {
      const userId = 'id';
      const setPassword = jest.fn(() => Promise.resolve('user'));
      password.setStore({ setPassword } as any);
      const passwordAuthenticator = jest
        .spyOn(password, 'passwordAuthenticator')
        .mockImplementation(() => Promise.resolve({}));
      await password.changePassword(userId, 'old-password', 'new-password');
      expect(passwordAuthenticator.mock.calls[0][0]).toEqual({ id: userId });
      expect(passwordAuthenticator.mock.calls[0][1]).toEqual('old-password');
      expect(setPassword.mock.calls[0][0]).toEqual(userId);
      expect(setPassword.mock.calls[0][1]).toBeTruthy();
      password.passwordAuthenticator.mockRestore();
    });
  });

  describe('sendVerificationEmail', () => {
    const email = 'john.doe@gmail.com';
    const verifiedEmail = 'john.doe2@gmail.com';
    const validUser = {
      emails: [{ address: email }, { address: verifiedEmail, verified: true }],
    };
    const invalidUser = {};

    it('throws if email is empty', async () => {
      try {
        await password.sendVerificationEmail('');
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('throws if user is not found', async () => {
      const findUserByEmail = jest.fn(() => Promise.resolve());
      password.setStore({ findUserByEmail } as any);
      try {
        await password.sendVerificationEmail(email);
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('throws when invalid address', async () => {
      const findUserByEmail = jest.fn(() => Promise.resolve(invalidUser));
      password.setStore({ findUserByEmail } as any);
      try {
        await password.sendVerificationEmail(email);
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('send email to first unverified email', async () => {
      const findUserByEmail = jest.fn(() => Promise.resolve(validUser));
      const addEmailVerificationToken = jest.fn(() => Promise.resolve());
      const prepareMail = jest.fn(() => Promise.resolve());
      const sanitizeUser = jest.fn(() => Promise.resolve());
      const sendMail = jest.fn(() => Promise.resolve());
      password.setStore({ findUserByEmail, addEmailVerificationToken } as any);
      password.server = {
        prepareMail,
        options: { sendMail },
        sanitizeUser,
      } as any;
      set(password.server, 'options.emailTemplates', {});
      await password.sendVerificationEmail(verifiedEmail);
      expect(addEmailVerificationToken.mock.calls[0].length).toBe(3);
      expect(prepareMail.mock.calls[0].length).toBe(6);
      expect(sendMail.mock.calls[0].length).toBe(1);
    });

    it('send email', async () => {
      const findUserByEmail = jest.fn(() => Promise.resolve(validUser));
      const addEmailVerificationToken = jest.fn(() => Promise.resolve());
      const prepareMail = jest.fn(() => Promise.resolve());
      const sanitizeUser = jest.fn(() => Promise.resolve());
      const sendMail = jest.fn(() => Promise.resolve());
      password.setStore({ findUserByEmail, addEmailVerificationToken } as any);
      password.server = {
        prepareMail,
        options: { sendMail },
        sanitizeUser,
      } as any;
      set(password.server, 'options.emailTemplates', {});
      await password.sendVerificationEmail(email);
      expect(addEmailVerificationToken.mock.calls[0].length).toBe(3);
      expect(prepareMail.mock.calls[0].length).toBe(6);
      expect(sendMail.mock.calls[0].length).toBe(1);
    });
  });

  describe('sendResetPasswordEmail', () => {
    const email = 'john.doe@gmail.com';
    const validUser = { emails: [{ address: email }] };

    it('throws if email is empty', async () => {
      try {
        await password.sendResetPasswordEmail('');
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('throws if user is not found', async () => {
      const findUserByEmail = jest.fn(() => Promise.resolve());
      password.setStore({ findUserByEmail } as any);
      try {
        await password.sendResetPasswordEmail(email);
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('send email', async () => {
      const findUserByEmail = jest.fn(() => Promise.resolve(validUser));
      const addResetPasswordToken = jest.fn(() => Promise.resolve());
      const prepareMail = jest.fn(() => Promise.resolve());
      const sanitizeUser = jest.fn(() => Promise.resolve());
      const sendMail = jest.fn(() => Promise.resolve());
      const getFirstUserEmail = jest.fn(() => Promise.resolve(email));
      password.setStore({ findUserByEmail, addResetPasswordToken } as any);
      password.server = {
        prepareMail,
        options: { sendMail },
        sanitizeUser,
        getFirstUserEmail,
      } as any;
      set(password.server, 'options.emailTemplates', {});
      await password.sendResetPasswordEmail(email);
      expect(addResetPasswordToken.mock.calls[0].length).toBe(4);
      expect(prepareMail.mock.calls[0].length).toBe(6);
      expect(sendMail.mock.calls[0].length).toBe(1);
    });
  });

  describe('sendEnrollmentEmail', () => {
    const email = 'john.doe@gmail.com';
    const validUser = { emails: [{ address: email }] };

    it('throws if user is not found', async () => {
      const findUserByEmail = jest.fn(() => Promise.resolve());
      password.setStore({ findUserByEmail } as any);
      try {
        await password.sendEnrollmentEmail(email);
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('send email', async () => {
      const findUserByEmail = jest.fn(() => Promise.resolve(validUser));
      const addResetPasswordToken = jest.fn(() => Promise.resolve());
      const prepareMail = jest.fn(() => Promise.resolve());
      const sanitizeUser = jest.fn(() => Promise.resolve());
      const sendMail = jest.fn(() => Promise.resolve());
      const getFirstUserEmail = jest.fn(() => Promise.resolve(email));
      password.setStore({ findUserByEmail, addResetPasswordToken } as any);
      password.server = {
        prepareMail,
        options: { sendMail },
        sanitizeUser,
        getFirstUserEmail,
      } as any;
      set(password.server, 'options.emailTemplates', {});
      await password.sendEnrollmentEmail(email);
      expect(addResetPasswordToken.mock.calls[0].length).toBe(4);
      expect(prepareMail.mock.calls[0].length).toBe(6);
      expect(sendMail.mock.calls[0].length).toBe(1);
    });
  });

  describe('createUser', async () => {
    it('throws on required fields', async () => {
      try {
        await password.createUser({
          password: '123456',
          username: '',
          email: '',
        });
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('throws when username already exists', async () => {
      const findUserByUsername = jest.fn(() => Promise.resolve('user'));
      password.setStore({ findUserByUsername } as any);
      try {
        await password.createUser({
          password: '123456',
          username: 'user1',
        });
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('throws when email already exists', async () => {
      const findUserByEmail = jest.fn(() => Promise.resolve('user'));
      password.setStore({ findUserByEmail } as any);
      try {
        await password.createUser({
          password: '123456',
          email: 'email1@email.com',
        });
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('throws if validateNewUser does not pass', async () => {
      const tmpAccountsPassword = new AccountsPassword({
        validateNewUser: () => Promise.resolve(false),
      });
      const findUserByEmail = jest.fn(() => Promise.resolve());
      tmpAccountsPassword.setStore({ findUserByEmail } as any);
      try {
        await tmpAccountsPassword.createUser({
          password: '123456',
          email: 'email1@email.com',
        });
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('create user', async () => {
      const findUserByEmail = jest.fn(() => Promise.resolve());
      const createUser = jest.fn(() => Promise.resolve());
      password.setStore({ findUserByEmail, createUser } as any);
      await password.createUser({
        password: '123456',
        email: 'email1@email.com',
      });
      expect(findUserByEmail.mock.calls.length).toBe(1);
      expect(createUser.mock.calls.length).toBe(1);
    });
  });
});
