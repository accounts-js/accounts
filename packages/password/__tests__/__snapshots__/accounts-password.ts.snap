// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsPassword addEmail call this.db.addEmail 1`] = `
Array [
  "id",
  "john.doe@gmail.com",
  true,
]
`;

exports[`AccountsPassword changePassword, invalidate all sessions and remove all reset password tokens from user storage call invalidateAllSessions 1`] = `
Array [
  undefined,
]
`;

exports[`AccountsPassword changePassword, invalidate all sessions and remove all reset password tokens from user storage call removeAllResetPasswordTokens 1`] = `
Array [
  undefined,
]
`;

exports[`AccountsPassword findUserByEmail call this.db.findUserByEmail 1`] = `
Array [
  "email",
]
`;

exports[`AccountsPassword findUserByUsername call this.db.findUserByUsername 1`] = `
Array [
  "email",
]
`;

exports[`AccountsPassword removeEmail call this.db.removeEmail 1`] = `
Array [
  "id",
  "email",
]
`;

exports[`AccountsPassword resetPassword reset password and invalidate all sessions 1`] = `
Array [
  "id",
]
`;

exports[`AccountsPassword resetPassword reset password and return tokens 1`] = `
Array [
  "id",
]
`;

exports[`AccountsPassword resetPassword validate user email if enrolled 1`] = `
Array [
  "id",
]
`;

exports[`AccountsPassword verifyEmail call this.db.verifyEmail 1`] = `
Array [
  undefined,
  "john.doe@gmail.com",
]
`;
